Identity.BusinessLogic/Services/Implementations/EShopAuthenticationService.cs

```diff
@@ -0,0 +1,98 @@
+    {
+        private readonly UserManager<User> _userManager = userManager;
+
+        private readonly ITokenService _tokenService = tokenService;
+        private readonly IMapper _mapper = mapper;
+        private readonly ILogger<EShopAuthenticationService> _logger = logger;
+
+        public async Task<TokenDTO?> AuthenticateAsync(LoginDTO loginCredentials)
+        {
+            User user = await ValidateAuthenticationAsync(loginCredentials.Email, loginCredentials.Password);
+
+            var tokenDto = await _tokenService.CreateTokenAsync(user, populateExp: true);
+
+            return tokenDto;
+        }
+
+        public async Task<bool> RegisterUserAsync(RegisterDTO registerCredentials)
+        {
+            await ValidateRegisterAsync(registerCredentials.Email, registerCredentials.Username);
+
+            var user = _mapper.Map<User>(registerCredentials);
+
+            await CreateUserAsync(user, registerCredentials.Password);
+
+            _logger.LogInformation("Adding role {role} to user {name}", Roles.User, user.UserName);
+            await _userManager.AddToRoleAsync(user, Roles.User);
+
+            return true;
+        }
+
+        private async Task<User> ValidateAuthenticationAsync(string email, string password)
+        {
+            _logger.LogInformation("Searching for user with email {email} ", email);
+            var user = await _userManager.FindByEmailAsync(email);
+
+            if (user is null)
+            {
+                _logger.LogDebug("No User");
+                throw new BadRequestException("User does not exist");
+            }
+
+            _logger.LogInformation("User: {id} {email} {name}", user.Id, user.Email, user.UserName);
+            var IsPasswordValid = await _userManager.CheckPasswordAsync(user, password);
+
+            if (!IsPasswordValid)
+            {
+                _logger.LogDebug("Invalid Password");
+                throw new BadRequestException("Invalid Password");
+            }
+
+            return user;
+        }
+
+        private async Task ValidateRegisterAsync(string email, string username)
+        {
+            _logger.LogDebug("Searching for user with email {email} ", email);
+            var emailTaken = await _userManager.FindByEmailAsync(email);
+
+            if (emailTaken is not null)
+            {
+                _logger.LogDebug("User found: {id} {name} {email}", emailTaken.Id, emailTaken.UserName, emailTaken.Email);
+                throw new BadRequestException("Email already taken");
+            }
+
+            var userNameTaken = await _userManager.FindByNameAsync(username);
+
+            if (userNameTaken is not null)
+            {
+                _logger.LogDebug("User found: {id} {name} {email}", userNameTaken.Id, userNameTaken.UserName, userNameTaken.Email);
+                throw new BadRequestException("Username already taken");
+            }
```

Comment:
you have already identified that such a user exists, why are you checking further?